### PR TITLE
Add support for noise IX, XX, and IK handshakes.

### PR DESCRIPTION
**Pull request by twittner**

### CODE REVIEW

File: protocols/noise/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+curve25519-dalek = "1"
+futures = "0.1"
+lazy_static = "1.2"
+libp2p-core = { version = "0.2.0", path = "../../core" }
+log = "0.4"
+rand = "0.6"
+tokio-io = "0.1"
+
+[dependencies.snow]
+git = "https://github.com/mcginty/snow.git"
```
Comment (tomaka) [inline protocols/noise/Cargo.toml]: Is the crates.io version lacking features?

File: protocols/noise/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    Noise(SnowError),
+    InvalidKey,
+    #[doc(hidden)]
+    __Nonexhaustive
+}
+
+impl fmt::Display for NoiseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            NoiseError::Io(e) => write!(f, "i/o: {:?}", e),
```
Comment (tomaka) [inline protocols/noise/src/error.rs]:
```suggestion
            NoiseError::Io(e) => write!(f, "{}", e),
```

File: protocols/noise/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            NoiseError::Noise(e) => write!(f, "noise: {:?}", e),
```
Comment (tomaka) [inline protocols/noise/src/error.rs]:
```suggestion
            NoiseError::Noise(e) => write!(f, "{}", e),
```

File: protocols/noise/src/keys.rs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    fn as_ref(&self) -> &[u8] {
+        &self.bytes
+    }
+}
+
+pub struct SecretKey {
+    scalar: Scalar
+}
+
+impl SecretKey {
```
Comment (tomaka) [inline protocols/noise/src/keys.rs]: Would be nice to eventually move this type in the core or something.
cc #807

File: examples/chat-noise.rs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+use std::io;
+use tokio::runtime::Runtime;
+
+#[derive(Debug, StructOpt)]
+struct Cli {
+    #[structopt(long = "dial", short = "d")]
+    dial: Option<String>
+}
+
+fn main() -> CliResult {
```
Comment (tomaka) [inline examples/chat-noise.rs]: IMO a `chat-noise` example doesn't really make sense. If you want to test whether it works, I suggest an example or a test in `libp2p-noise`.
In the same vein, we're not going to create `chat-mplex`, `chat-yamux`, `chat-quic`, etc.
Integrating noise in the unique `chat` example does make sense to me however.

File: protocols/noise/Cargo.toml (hunk 1 above)
Comment (twittner) [reply protocols/noise/Cargo.toml]: There was a slight inconvenience related to their `Random` trait which was removed on `master`. I could switch to the crates.io version with minor effort or do it before we publish next time.

**Comment by burdges** [comment]
I brought this up recently in https://forum.web3.foundation/t/transport-layer-authentication-libp2ps-secio/69  We might eventually switch to  https://noiseexplorer.com/  if it gets Rust support.

**Comment by tomaka** [comment]
Note that we need #882 before this is usable in practice.

File: protocols/noise/Cargo.toml (hunk 1 above)
Comment (twittner) [reply protocols/noise/Cargo.toml]: Version `0.5.0-alpha1` was just released to crates.io. The PR has been updated to use this version.

File: protocols/noise/src/xx.rs [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+            match mem::replace(&mut self.0, InboundState::Done) {
+                InboundState::Init(io, config) => {
+                    let session = snow::Builder::with_resolver(config.params, Box::new(Resolver))
+                        .local_private_key(config.keypair.secret().as_ref())
+                        .build_responder()?;
+                    let output = NoiseOutput::new(io, session);
+                    self.0 = InboundState::RecvHandshake1(output)
+                }
+                InboundState::RecvHandshake1(mut io) => {
+                    if io.poll_read(&mut [])?.is_ready() {
```
Comment (tomaka) [inline protocols/noise/src/xx.rs]: Every single call to `poll_read` or `poll_write` passes an empty slice.
What's the reason for going for `AsyncRead` and `AsyncWrite` if you're always passing empty slices?

Comment (twittner) [reply protocols/noise/src/xx.rs]: During the handshake phase there is no application data to send, so effectively only key material is exchanged. The trace log highlights what is happening. Here is an excerpt for an initial XX handshake write:

```
[TRACE libp2p_noise::io] write state: Init
[TRACE libp2p_noise::io] write state: BufferData { off: 0 }
[TRACE libp2p_noise::io] write: buffered 0 bytes
[TRACE libp2p_noise::io] flush: encrypting 0 bytes
[TRACE libp2p_noise::io] flush: cipher text len = 32 bytes
[TRACE libp2p_noise::io] flush: writing len (32)
[TRACE libp2p_noise::io] flush: wrote 32/32 bytes
[TRACE libp2p_noise::io] flush: finished writing 32 bytes
...
```

Comment (tomaka) [reply protocols/noise/src/xx.rs]: Oh I see, the `Read` and `Write` are used post-handshake.

File: protocols/noise/src/util.rs [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        let mut s = [0; 32];
+        rng.fill_bytes(&mut s);
+        let secret = SecretKey::new(s);
+        let public = secret.public();
+        self.keypair = Keypair::new(secret, public)
+    }
+
+    fn dh(&self, pubkey: &[u8], out: &mut [u8]) -> Result<(), ()> {
+        let mut p = [0; 32];
+        p.copy_from_slice(&pubkey[.. 32]);
```
Comment (tomaka) [inline protocols/noise/src/util.rs]: Here you do `&pubkey[..32]` while in `set` you do just `privkey`. I'd prefer to add a verification that the lengths of `pubkey` and `out` are `32`.

File: Cargo.toml [hunk 8]
```diff
@@ -38,14 +38,17 @@ tokio-io = "0.1"
...
 libp2p-mdns = { version = "0.2.0", path = "./misc/mdns" }
+libp2p-noise = { version = "0.1.0", path = "./protocols/noise" }
 libp2p-tcp = { version = "0.2.0", path = "./transports/tcp" }
 
 [target.'cfg(any(target_os = "emscripten", target_os = "unknown"))'.dependencies]
 stdweb = { version = "0.4", default-features = false }
 
 [dev-dependencies]
 env_logger = "0.6.0"
+quicli = "0.4"
```
Comment (tomaka) [inline Cargo.toml]: I suppose that this and `structopt` are no longer needed anymore, since the example is gone.

File: protocols/noise/src/io.rs [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+const_assert_eq!(A; READ_END - READ_BEGIN, MAX_NOISE_PKG_LEN);
+const_assert_eq!(B; READ_CRYPTO_END - READ_CRYPTO_BEGIN, MAX_NOISE_PKG_LEN);
+const_assert_eq!(C; WRITE_END - WRITE_BEGIN, MAX_WRITE_BUF_LEN);
+const_assert_eq!(D; WRITE_CRYPTO_END - WRITE_CRYPTO_BEGIN, 2 * MAX_WRITE_BUF_LEN);
+const_assert_eq!(E; 180222, TOTAL_BUFFER_LEN);
+const_assert_eq!(F; 0, READ_CRYPTO_BEGIN - READ_END); // read crypto buffer follows read buffer
+const_assert_eq!(G; 0, WRITE_CRYPTO_BEGIN - WRITE_END); // write crypto buffer follows write buffer
+
+pub struct NoiseOutput<T> {
+    pub(super) io: T,
```
Comment (tomaka) [inline protocols/noise/src/io.rs]: I'm not a fan of adding `super` everywhere. I think it's better to keep everything internal to the module. If you ever need to debug this, it's easier to know that all the modifications to the state are done within the module.

File: protocols/noise/src/io.rs [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+    pub(super) session: snow::Session,
+    pub(super) buffer: Box<[u8; TOTAL_BUFFER_LEN]>,
```
Comment (tomaka) [inline protocols/noise/src/io.rs]: An alternative way than `split_at_mut` could do to create a separate type:

```rust
struct Buffer {
    inner: Box<[u8; TOTAL_BUFFER_LEN]>
}

impl Buffer {
     fn borrow(&mut self) -> BufferBorrow { /* call split_at_mut here */ }
}

struct BufferBorrow<'a> {
    read: &'a mut [u8],
    read_crypto: &'a mut [u8],
    write: &'a mut [u8],
    write_crypto: &'a mut [u8],
}
```

File: protocols/noise/src/io.rs [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+
+const_assert_eq!(A; READ_END - READ_BEGIN, MAX_NOISE_PKG_LEN);
+const_assert_eq!(B; READ_CRYPTO_END - READ_CRYPTO_BEGIN, MAX_NOISE_PKG_LEN);
+const_assert_eq!(C; WRITE_END - WRITE_BEGIN, MAX_WRITE_BUF_LEN);
+const_assert_eq!(D; WRITE_CRYPTO_END - WRITE_CRYPTO_BEGIN, 2 * MAX_WRITE_BUF_LEN);
+const_assert_eq!(E; 180222, TOTAL_BUFFER_LEN);
+const_assert_eq!(F; 0, READ_CRYPTO_BEGIN - READ_END); // read crypto buffer follows read buffer
+const_assert_eq!(G; 0, WRITE_CRYPTO_BEGIN - WRITE_END); // write crypto buffer follows write buffer
+
+pub struct NoiseOutput<T> {
```
Comment (tomaka) [inline protocols/noise/src/io.rs]: I'm not a fan of calling `poll_read` and `poll_write` with empty slices. It's a bit too weird.
I suggest we keep `NoiseOutput` as the "final" type (the one that the user receives), but add a `Handshake` type that wraps around `NoiseOutput`, performs the handshake, and has a `into_transport_mode()` method that turns it into a `NoiseOutput`. This way you can also make `session` private.

File: protocols/noise/src/util.rs (hunk 7 above)
Comment (twittner) [reply protocols/noise/src/util.rs]: The `pubkey` that is passed in is longer than 32 bytes, hence the need to restrict it. When called from [`HandshakeState::dh`](https://github.com/mcginty/snow/blob/953deb8f09c4f0908d4c820f61e9a4a84c6ac1ae/src/handshakestate.rs#L136) it uses its `re` or `rs` slices which are 56 bytes each.

**Pull request merged**